src/middleware/context-storage/index.ts

```diff
@@ -0,0 +1,55 @@
+/**
+ * @module
+ * Context Storage Middleware for Hono.
+ */
+
+import type { Context } from '../../context'
+import type { Env, MiddlewareHandler } from '../../types'
+import { AsyncLocalStorage } from 'node:async_hooks'
+
+const asyncLocalStorage = new AsyncLocalStorage<Context>()
+
+/**
+ * Context Storage Middleware for Hono.
+ *
+ * @see {@link https://hono.dev/docs/middleware/builtin/context-storage}
+ *
+ * @returns {MiddlewareHandler} The middleware handler function.
+ *
+ * @example
+ * ```ts
+ * type Env = {
+ *   Variables: {
+ *     message: string
+ *   }
+ * }
+ *
+ * const app = new Hono<Env>()
+ *
+ * app.use(contextStorage())
+ *
+ * app.use(async (c, next) => {
+ *   c.set('message', 'Hono is cool!!)
+ *   await next()
+ * })
+ *
+ * app.get('/', async (c) => { c.text(getMessage()) })
+ *
+ * const getMessage = () => {
+ *   return getContext<Env>().var.message
+ * }
+ * ```
+ */
+export const contextStorage = (): MiddlewareHandler => {
+  return async function contextStorage(c, next) {
+    await asyncLocalStorage.run(c, next)
+  }
+}
+
+export const getContext = <E extends Env = Env>() => {
+  const context = asyncLocalStorage.getStore() as Context<E>
```

Comment:
This is a personal taste. I like the following way to enable the recognition of type definitions with the definition of `getContext`.

```diff
diff --git a/src/middleware/context-storage/index.ts b/src/middleware/context-storage/index.ts
index 8636f075..2545d3ca 100644
--- a/src/middleware/context-storage/index.ts
+++ b/src/middleware/context-storage/index.ts
@@ -46,8 +46,8 @@ export const contextStorage = (): MiddlewareHandler => {
   }
 }

-export const getContext = <E extends Env = Env>() => {
-  const context = asyncLocalStorage.getStore() as Context<E>
+export const getContext = <E extends Env = Env>(): Context<E> => {
+  const context = asyncLocalStorage.getStore()
   if (!context) {
     throw new Error('Context is not available')
   }
```